test/distributed/elastic/rendezvous/utils_test.py

```diff
@@ -215,3 +220,103 @@ def test_matches_machine_hostname_returns_false_if_hostname_does_not_match(self)
         for host in hosts:
             with self.subTest(host=host):
                 self.assertFalse(_matches_machine_hostname(host))
+
+
+class PeriodicTimerTest(TestCase):
+    def test_start_can_be_called_only_once(self):
+        timer = _PeriodicTimer(timedelta(seconds=1), lambda: None)
+
+        timer.start()
+
+        with self.assertRaisesRegex(RuntimeError, r"^The timer has already started.$"):
+            timer.start()
+
+        timer.cancel()
+
+    def test_cancel_can_be_called_multiple_times(self):
+        timer = _PeriodicTimer(timedelta(seconds=1), lambda: None)
+
+        timer.start()
+
+        timer.cancel()
+        timer.cancel()
+
+    def test_cancel_stops_background_thread(self):
+        timer = _PeriodicTimer(timedelta(seconds=1), lambda: None)
+
+        timer.start()
+
+        self.assertTrue(any(t.name == "PeriodicTimer" for t in threading.enumerate()))
+
+        timer.cancel()
+
+        self.assertTrue(all(t.name != "PeriodicTimer" for t in threading.enumerate()))
+
+    def test_delete_stops_background_thread(self):
```

Comment:
Nit: you can probably remove the duplicate lines for this and the test above by making a helper function and using a flag to call `del` or `cancel`

Reply:
In fact I thought about that; the tricky thing is that `del` is scope bound. Calling `del` in a helper function will only decrease the refcount of the timer back to 1 because we have to pass it to that function either as an argument or as part of the closure which will implicitly increase its refcount. In order to make sure that `del` really disposes the instance (meaning its refcount becomes 0), I have to make sure that there are no dangling references to it. I might have probably come up with a clever trick if I spent more time on it, but I was just too lazy and decided it is not worth the trouble for only 3-4 lines of code.